File: pyatlas/pyatlas/edge.py
import node
import atlas_entity
import entity_type
import identifier_converters


class Edge(atlas_entity.AtlasEntity):
    """
    A unidirectional Atlas Edge. Bidirectional OSM ways are represented with
    two opposing Edges, where one of them is the master Edge. The master Edge
    will have a positive identifier and the same traffic direction as OSM.
    """

    def __init__(self, parent_atlas, index):
        """
        Constuct a new Edge. This should not be called directly.
        """
        super(Edge, self).__init__(parent_atlas)
        self.index = index

    def __str__(self):
        """
        Get a string representation of this Edge.
        """
        result = '[ '
        result += 'Edge: id=' + str(self.get_identifier())

        string = ""
        string += str(self.get_start().get_identifier()) + ','
        result += ', start=[' + string + ']'

        string = ""
        string += str(self.get_end().get_identifier()) + ','
        result += ', end=[' + string + ']'

        result += ', geom=' + str(self.as_polyline())
        result += ', tags=' + str(self.get_tags())

        string = ''
        for relation in self.get_relations():
            string += str(relation.get_identifier()) + ','
        result += ', relations=[' + string + ']'

        result += ' ]'
        return result

    def __lt__(self, other):
        """
        Custom implementation of less-than so that collections of Edges can be
        easily sorted.
        """
        return self.get_identifier() < other.get_identifier()

    def as_polyline(self):
        """
        Get the PolyLine geometry of this Edge.
        """
        return self.get_parent_atlas()._get_edgePolyLines()[self.index]

    def get_identifier(self):
        """
        Get the Atlas identifier of this Edge.
        :return: the Atlas id
        """
        return self.get_parent_atlas()._get_edgeIdentifiers().elements[self.index]

    def get_tags(self):
        """
        Get a dictionary of this Edge's tags.
        """
        edge_tag_store = self.get_parent_atlas()._get_edgeTags()
        return edge_tag_store.to_key_value_dict(self.index)

    def get_bounds(self):
        """
        Get the bounding Rectangle of this Edge.
        """
        return self.as_polyline().get_bounds()

    def get_relations(self):
        """
        Get the frozenset of Relations of which this Edge is a member.
        Returns an empty set if this Edge is not a member of any Relations.
        """
        relation_map = self.get_parent_atlas()._get_edgeIndexToRelationIndices()
        return self._get_relations_helper(relation_map, self.index)

    def get_connected_nodes(self):
        """
        Get a frozenset of the Nodes connected to this Edge.
        """
        result = set()
        result.add(self.get_start())
        result.add(self.get_end())
        return frozenset(result)

    def get_connected_edges(self):
        """
        Get a frozenset of the Edges connected at the ends of the Nodes of this
        Edge. The set will not contain the Edge this is method called on, but
        will contain the reversed Edge if this Edge is part of a two-way road.
        """
        result = set()
        for edge0 in self.get_end().get_connected_edges():
            if self != edge0:
                result.add(edge0)
        for edge0 in self.get_start().get_connected_edges():
            if self != edge0:
                result.add(edge0)
        return result

    def get_start(self):
        """
        Get the starting Node of this Edge.
        """
        edge_start_node_index = self.get_parent_atlas()._get_edgeStartNodeIndex()
        index = edge_start_node_index.elements[self.index]
        return node.Node(self.get_parent_atlas(), index)

    def get_end(self):
        """
        Get the ending Node of this Edge.
        """
        edge_end_node_index = self.get_parent_atlas()._get_edgeEndNodeIndex()
        index = edge_end_node_index.elements[self.index]
        return node.Node(self.get_parent_atlas(), index)

    def get_master_edge(self):
        """
        Get the master for this Edge. Returns itself if this is the master Edge.
        """
        if self.is_master_edge():
            return self
        else:
            return self.get_parent_atlas().edge(-1 * self.get_identifier())

    def is_master_edge(self):
        """
        Checks if this Edge is a master edge.
        """
        return self.get_identifier() > 0

    def has_reversed_edge(self):
        """
        Checks if this Edge is a member of a bidirectional Edge pairing.
        """
        return self.get_parent_atlas().edge(-1 * self.get_identifier()) is not None

    def is_reversed_edge(self, candidate):
        """
        Check if the candidate Edge is the bidirectional reverse of this Edge.
        """
        if candidate.get_type() != entity_type.EntityType.EDGE:
            return False
        return self.get_identifier() == -1 * candidate.get_identifier()

    def get_reversed_edge(self):
        """
        Get the bidirectional pair Edge to this Edge, if it exists. Returns None
        if it does not.
        """
        if not self.has_reversed_edge():
            return None
        return self.get_parent_atlas().edge(-1 * self.get_identifier())

    def get_highway_tag_value(self):
        """
        Get the value of the "highway" tag of this Edge, if present. Returns
        None if there is no "highway" tag.
        """
        tags = self.get_tags()
        if 'highway' in tags:
            return tags['highway']
        else:
            return None

    def is_connected_at_end_to(self, candidates):
        """
        Given a set of AtlasEntity candidates, test if this edge is directly
        connected at its end to at least one of the candidates.
        """
        for entity in candidates:
            if entity.get_type() == entity_type.EntityType.NODE:
                if self.get_end() == entity:
                    return True
            if entity.get_type() == entity_type.EntityType.EDGE:
                if self.get_end() == entity.get_start():
                    return True
        return False

    def is_connected_at_start_to(self, candidates):
        """
        Given a set of AtlasEntity candidates, test if this edge is directly
        connected at its start to at least one of the candidates.
        """
        for entity in candidates:
            if entity.get_type() == entity_type.EntityType.NODE:
                if self.get_start() == entity:
                    return True
            if entity.get_type() == entity_type.EntityType.EDGE:
                if self.get_start() == entity.get_end():
                    return True
        return False

    def is_way_sectioned(self):
        """
        Determine if this Edge is a way-sectioned road.
        """
        return identifier_converters.get_way_section_index(self.get_identifier()) != 0

    def get_type(self):
        """
        Implement superclass get_type(). Always returns EntityType.EDGE.
        """
        return entity_type.EntityType.EDGE

File: src/main/java/org/openstreetmap/atlas/geography/atlas/items/Edge.java
package org.openstreetmap.atlas.geography.atlas.items;

import java.util.HashSet;
import java.util.Optional;
import java.util.Set;

import org.apache.commons.lang3.Validate;
import org.openstreetmap.atlas.exception.CoreException;
import org.openstreetmap.atlas.geography.atlas.Atlas;
import org.openstreetmap.atlas.geography.atlas.pbf.slicing.identifier.ReverseIdentifierFactory;
import org.openstreetmap.atlas.tags.HighwayTag;

import com.google.gson.JsonObject;

/**
 * A unidirectional edge that belongs to an Atlas.
 *
 * @author matthieun
 */
public abstract class Edge extends LineItem implements Comparable<Edge>
{
    private static final long serialVersionUID = -4426003484206550921L;
    private static final ReverseIdentifierFactory reverseIdentifierFactory = new ReverseIdentifierFactory();

    /**
     * If the way is bidirectional in OSM, we will put two edges in atlas for two traffic
     * directions, each of them is one way. The main edge will have positive identifier and same
     * traffic direction as OSM.
     *
     * @param identifier
     *            Edge identifier
     * @return True if the edge identifier is positive
     */
    public static boolean isMainEdgeIdentifier(final long identifier)
    {
        return identifier > 0;
    }

    /**
     * @param identifier
     *            Edge identifier
     * @return True if the edge identifier is positive
     * @deprecated Use isMainEdgeIdentifier instead.
     */
    @Deprecated(since = "")
    public static boolean isMasterEdgeIdentifier(final long identifier)
    {
        return isMainEdgeIdentifier(identifier);
    }

    protected Edge(final Atlas atlas)
    {
        super(atlas);
    }

    /**
     * Compare two edges on their identifier.
     * <p>
     * NOSONAR here as the {@link AtlasEntity} equals and hashcode are good enough. ""equals(Object
     * obj)" should be overridden along with the "compareTo(T obj)" method (squid:S1210)"
     */
    @Override
    public int compareTo(final Edge other) // NOSONAR
    {
        return Long.compare(this.getIdentifier(), other.getIdentifier());
    }

    /**
     * @return All the {@link Edge}s connected to the end {@link Node}s of this {@link Edge}, except
     *         self. If this {@link Edge} is a two-way road, then the reversed {@link Edge} will be
     *         included in the set.
     */
    public Set<Edge> connectedEdges()
    {
        final Set<Edge> result = new HashSet<>();
        for (final Edge edge : this.end().connectedEdges())
        {
            if (!this.equals(edge))
            {
                result.add(edge);
            }
        }
        for (final Edge edge : this.start().connectedEdges())
        {
            if (!this.equals(edge))
            {
                result.add(edge);
            }
        }
        return result;
    }

    public Node connectedNode(final ConnectedNodeType connectedNodeType)
    {
        Validate.notNull(connectedNodeType);
        final Node connectedNode = connectedNodeType.getAccessFunction().apply(this);
        return connectedNode;
    }

    public Set<Node> connectedNodes()
    {
        final Set<Node> result = new HashSet<>();
        result.add(this.start());
        result.add(this.end());
        return result;
    }

    /**
     * @return The same {@link Edge} but with the tags interpreted with this {@link Edge}'s
     *         direction. For example, if this {@link Edge} is backwards from its OSM way, and the
     *         way has a maxspeed:backward tag, here it will be translated into a maxspeed tag. Also
     *         the maxspeed:forward tag will be filtered out (it will be used by the reverse edge).
     */
    public Edge directionalized()
    {
        return new DirectionalizedEdge(this);
    }

    /**
     * @return The {@link Node} at the end of this {@link Edge}
     */
    public abstract Node end();

    @Override
    public JsonObject getGeoJsonProperties()
    {
        final JsonObject properties = super.getGeoJsonProperties();

        properties.addProperty(ConnectedNodeType.START.getPropertyName(), start().getIdentifier());
        properties.addProperty(ConnectedNodeType.END.getPropertyName(), end().getIdentifier());

        return properties;
    }

    /**
     * @return the main for this {@link Edge}, which may or may not be the main.
     */
    public Edge getMainEdge()
    {
        return this.isMainEdge() ? this
                : this.reversed()
                        .orElseThrow(() -> new CoreException(
                                "Reverse edge should be available for edge {}",
                                this.getIdentifier()));
    }

    public long getMainEdgeIdentifier()
    {
        return Math.abs(this.getIdentifier());
    }

    /**
     * @return the main for this {@link Edge}, which may or may not be the main.
     * @deprecated Use getMainEdge instead.
     */
    @Deprecated(since = "")
    public Edge getMasterEdge()
    {
        return getMainEdge();
    }

    /**
     * @return The main edge identifier
     * @deprecated Use getMainEdgeIdentifier instead
     */
    @Deprecated(since = "")
    public long getMasterEdgeIdentifier()
    {
        return getMainEdgeIdentifier();
    }

    @Override
    public ItemType getType()
    {
        return ItemType.EDGE;
    }

    /**
     * @return {@code true} if there is a reverse edge to this one
     */
    public boolean hasReverseEdge()
    {
        return this.getAtlas().edge(-this.getIdentifier()) != null;
    }

    /**
     * @return The {@link HighwayTag} of the Edge, if it is present. Return HighwayTag.NO if it is
     *         not.
     */
    public HighwayTag highwayTag()
    {
        final Optional<HighwayTag> result = HighwayTag.highwayTag(this);
        if (result.isPresent())
        {
            return result.get();
        }
        else
        {
            return HighwayTag.NO;
        }
    }

    /**
     * @return All the {@link Edge}s connected and pointing to the start {@link Node} of this
     *         {@link Edge}. If this {@link Edge} is a two-way road, then the reversed {@link Edge}
     *         will be included in the set.
     */
    public Set<Edge> inEdges()
    {
        return this.start().inEdges();
    }

    /**
     * @param candidates
     *            Set of edges and nodes to test connectivity to.
     * @return True if the edge is directly connected at its end to at least one of the candidate
     *         items
     */
    public boolean isConnectedAtEndTo(final Set<? extends AtlasItem> candidates)
    {
        for (final AtlasItem item : candidates)
        {
            if (item instanceof Node && end().equals(item))
            {
                return true;
            }
            if (item instanceof Edge && end().equals(((Edge) item).start()))
            {
                return true;
            }
        }
        return false;
    }

    /**
     * @param candidates
     *            Set of edges and nodes to test connectivity to.
     * @return True if the edge is directly connected at its start to at least one of the candidate
     *         items
     */
    public boolean isConnectedAtStartTo(final Set<? extends AtlasItem> candidates)
    {
        for (final AtlasItem item : candidates)
        {
            if (item instanceof Node && start().equals(item))
            {
                return true;
            }
            if (item instanceof Edge && start().equals(((Edge) item).end()))
            {
                return true;
            }
        }
        return false;
    }

    /**
     * Checks if edge is a main edge, by verifying that its identifier is a main edge identifier.
     *
     * @return True if the edge's identifier is a main edge identifier
     */
    public boolean isMainEdge()
    {
        return isMainEdgeIdentifier(this.getIdentifier());
    }

    /**
     * @return True if the edge's identifier is a main edge identifier
     * @deprecated Use isMainEdge instead.
     */
    @Deprecated(since = "")
    public boolean isMasterEdge()
    {
        return isMainEdge();
    }

    /**
     * @param candidate
     *            candidate for reverseEdge
     * @return {@code true} if candidate is the reverse of this {@link Edge}
     */
    public boolean isReversedEdge(final Edge candidate)
    {
        return this.getIdentifier() == -candidate.getIdentifier();
    }

    /**
     * @return {@code true} if the {@link Edge} is a way-sectioned road.
     */
    public boolean isWaySectioned()
    {
        return reverseIdentifierFactory.getWaySectionIndex(this.getIdentifier()) != 0;
    }

    /**
     * @return All the {@link Edge}s connected and pointing out of the end {@link Node} of this
     *         {@link Edge}. If this {@link Edge} is a two-way road, then the reversed {@link Edge}
     *         will be included in the set.
     */
    public Set<Edge> outEdges()
    {
        return this.end().outEdges();
    }

    /**
     * @return An {@link Edge} that is reversed to this one if it exists, empty otherwise.
     */
    public Optional<Edge> reversed()
    {
        final Edge edge = this.getAtlas().edge(-this.getIdentifier());
        if (edge != null)
        {
            return Optional.of(edge);
        }
        return Optional.empty();
    }

    public abstract Node start();

    @Override
    public String toDiffViewFriendlyString()
    {
        final String relationsString = this.parentRelationsAsDiffViewFriendlyString();

        final String startNodeString = start() != null ? Long.toString(start().getIdentifier())
                : "null";
        final String endNodeString = start() != null ? Long.toString(end().getIdentifier())
                : "null";
        final String polyLineWkt = this.asPolyLine() != null ? this.asPolyLine().toWkt() : "null";

        return "[Edge" + ": id=" + this.getIdentifier() + ", startNode=" + startNodeString
                + ", endNode=" + endNodeString + ", polyLine=" + polyLineWkt + ", relations=("
                + relationsString + "), " + tagString() + "]";
    }

    @Override
    public String toString()
    {
        return "[Edge" + ": id=" + this.getIdentifier() + ", startNode=" + start().getIdentifier()
                + ", endNode=" + end().getIdentifier() + ", polyLine=" + this.asPolyLine().toWkt()
                + ", " + tagString() + "]";
    }
}
